File: auntieos-admin/src/lib/coveragePackage.test.ts
import { describe, expect, it } from 'vitest';
import {
  DEFAULT_COVERAGE_RULES,
  DEFAULT_DURATIONS,
  alignPinnedToDurations,
  buildDayPatterns,
  coverageForDay,
  daysBetween,
  durationsFromServiceRates,
  gapWarnings,
  minutesToTime,
  normalizePackage,
  priceDay,
  pricePackage,
  quoteText,
  timeToMinutes,
  todayIso,
  visitsFromPattern,
  withKind,
  type Duration,
  type Package,
} from './coveragePackage';

const DURATIONS: readonly Duration[] = DEFAULT_DURATIONS;
const OVERNIGHT = DURATIONS.find((d) => d.kind === 'overnight')!; // d7, $150, 720min

function pkg(over: Partial<Package> = {}): Package {
  return normalizePackage({ id: 'p1', name: 'Test', ...over });
}

describe('daysBetween', () => {
  it('counts inclusively, 0 for unset/reversed', () => {
    expect(daysBetween('2026-07-01', '2026-07-05')).toBe(5);
    expect(daysBetween('2026-07-05', '2026-07-01')).toBe(0);
    expect(daysBetween('', '2026-07-05')).toBe(0);
  });
});

describe('time helpers', () => {
  it('parses/formats', () => {
    expect(timeToMinutes('07:30')).toBe(450);
    expect(timeToMinutes('nope')).toBeNull();
    expect(minutesToTime(450)).toBe('7:30 AM');
    expect(minutesToTime(1320)).toBe('10:00 PM');
  });
});

describe('todayIso', () => {
  it('is the LOCAL calendar date, not the UTC one', () => {
    // 10:30 PM local on the 9th. `toISOString().slice(0, 10)` gives the 10th
    // anywhere west of Greenwich, which is the bug this helper exists to avoid.
    expect(todayIso(new Date(2026, 8, 9, 22, 30))).toBe('2026-09-09');
    expect(todayIso(new Date(2026, 0, 5, 0, 15))).toBe('2026-01-05');
  });
});

describe('durationsFromServiceRates (issue #693)', () => {
  const RATES = { '30Minute': '25', 'Half-Day 6Hrs': '100', Overnight: '150', Consultation: '' };

  it('reads the operator KinCare types as the visit menu, keyed by name', () => {
    const menu = durationsFromServiceRates(RATES);
    const byId = Object.fromEntries(menu.map((d) => [d.id, d]));
    expect(byId['30Minute']).toEqual({ id: '30Minute', label: '30Minute', minutes: 30, price: 25, kind: 'visit' });
    expect(byId['Half-Day 6Hrs']!.minutes).toBe(360);
    expect(byId['Half-Day 6Hrs']!.price).toBe(100);
  });

  it('reads a type named overnight as an overnight, and everything else as a visit', () => {
    const menu = durationsFromServiceRates(RATES);
    expect(menu.find((d) => d.id === 'Overnight')!.kind).toBe('overnight');
    // A 6-hour day stay is long, but it is still a visit: length is not the signal.
    expect(menu.find((d) => d.id === 'Half-Day 6Hrs')!.kind).toBe('visit');
  });

  it('takes a stated serviceDurations minute count over the length in the name', () => {
    const menu = durationsFromServiceRates({ Overnight: '150' }, { Overnight: '720' });
    expect(menu[0]!.minutes).toBe(720);
  });

  it('keeps a type with no price or no stated length, at 0', () => {
    const menu = durationsFromServiceRates(RATES);
    const consult = menu.find((d) => d.id === 'Consultation')!;
    expect(consult.price).toBe(0);
    expect(consult.minutes).toBe(0);
  });
});

describe('alignPinnedToDurations (issue #693)', () => {
  it('repoints a pinned visit whose length is not in the menu', () => {
    const menu = durationsFromServiceRates({ '30Minute': '25', Overnight: '150' });
    // The shipped default pins `d2`, an id from the deleted builder-owned menu.
    const aligned = alignPinnedToDurations(DEFAULT_COVERAGE_RULES, menu);
    expect(aligned.pinnedTimes[0]!.durationId).toBe('30Minute');
    // Only the id moves; the client's own label and time are untouched.
    expect(aligned.pinnedTimes[0]!.label).toBe(DEFAULT_COVERAGE_RULES.pinnedTimes[0]!.label);
    expect(aligned.pinnedTimes[0]!.time).toBe(DEFAULT_COVERAGE_RULES.pinnedTimes[0]!.time);
  });

  it('leaves a pinned visit that already points at a real length alone', () => {
    const menu = durationsFromServiceRates({ '30Minute': '25', '60Minute': '45' });
    const rules = { ...DEFAULT_COVERAGE_RULES, pinnedTimes: [{ id: 'p', label: 'Meds', time: '12:00', durationId: '60Minute' }] };
    expect(alignPinnedToDurations(rules, menu).pinnedTimes[0]!.durationId).toBe('60Minute');
  });

  it('blanks the length when the menu holds no visit type at all', () => {
    expect(alignPinnedToDurations(DEFAULT_COVERAGE_RULES, []).pinnedTimes[0]!.durationId).toBe('');
  });
});

describe('withKind', () => {
  it('defaults visits and migrates the legacy overnight id', () => {
    const [visit, legacy] = withKind([
      { id: 'x', label: 'A', minutes: 30, price: 20 },
      { id: 'd7', label: 'Overnight', minutes: 720, price: 150 },
    ]);
    expect(visit!.kind).toBe('visit');
    expect(legacy!.kind).toBe('overnight');
  });
});

describe('buildDayPatterns (suggestions)', () => {
  it('is empty for a reversed/degenerate window', () => {
    expect(buildDayPatterns(DURATIONS, [], 6, '14:00', '11:00')).toEqual([]);
    expect(buildDayPatterns(DURATIONS, [], 6, '11:00', '14:00')).toEqual([]); // 3h ≤ 6h gap, no pinned
  });

  it('only fills with visit-kind durations, never overnights', () => {
    const patterns = buildDayPatterns(DURATIONS, [], 3, '07:00', '22:00');
    expect(patterns.length).toBeGreaterThan(0);
    for (const p of patterns) {
      for (const tp of p.touchpoints) {
        expect(DURATIONS.find((d) => d.id === tp.durationId)!.kind).toBe('visit');
      }
    }
  });

  it('never leaves a gap wider than the max', () => {
    const patterns = buildDayPatterns(DURATIONS, [], 4, '07:00', '22:00');
    const maxGapMin = 4 * 60;
    for (const p of patterns) {
      const times = [420, ...p.touchpoints.map((t) => t.time), 1320].sort((a, b) => a - b);
      for (let i = 0; i < times.length - 1; i++) {
        expect(times[i + 1]! - times[i]!).toBeLessThanOrEqual(maxGapMin + 0.001);
      }
    }
  });

  it('seeds a package template from a suggestion', () => {
    const p = buildDayPatterns(DURATIONS, [{ id: 'p', label: 'Meds', time: '12:00', durationId: 'd3' }], 12, '07:00', '22:00')[0]!;
    const visits = visitsFromPattern(p);
    expect(visits.length).toBe(p.touchpoints.length);
    expect(visits[0]!.time).toBe(720);
  });
});

describe('coverageForDay + priceDay', () => {
  it('prices bare visits at list price with no overnight', () => {
    const { total, items } = priceDay(
      [{ id: 'v1', time: 720, durationId: 'd3', label: 'Lunch' }],
      DURATIONS,
      { eveningFrom: null, morningUntil: null, bonusFreeVisit: false },
    );
    expect(total).toBe(35); // d3 = $35
    expect(items[0]!.free).toBe(false);
  });

  it('a visit inside the overnight evening window is covered (free)', () => {
    // overnight tonight from 21:00, 2h buffer → covers from 19:00 onward.
    const p = pkg({ overnightStart: '21:00', overnightBufferHours: 2, overnightNights: { 0: true } });
    const cov = coverageForDay(p, 0, 2, OVERNIGHT);
    expect(cov.eveningFrom).toBe(19 * 60);
    const { items } = priceDay([{ id: 'v', time: 20 * 60, durationId: 'd3', label: 'Evening' }], DURATIONS, cov);
    expect(items[0]!.free).toBe(true);
    expect(items[0]!.price).toBe(0);
  });

  it('the morning after an overnight grants one free bonus visit', () => {
    // Prior night had an overnight → bonusFreeVisit true; the first paid visit goes free.
    const cov = { eveningFrom: null, morningUntil: null, bonusFreeVisit: true };
    const { items, total } = priceDay(
      [
        { id: 'a', time: 8 * 60, durationId: 'd3', label: 'AM' },
        { id: 'b', time: 12 * 60, durationId: 'd3', label: 'Noon' },
      ],
      DURATIONS,
      cov,
    );
    expect(items[0]!.bonus).toBe(true);
    expect(items[0]!.price).toBe(0);
    expect(total).toBe(35); // only the second visit is billed
  });
});

describe('pricePackage', () => {
  it('sums per-day costs incl. overnight and applies a discount', () => {
    const p = pkg({
      visits: [{ id: 'v', time: 12 * 60, durationId: 'd3', label: 'Lunch' }], // $35/day
      overnightNights: { 0: true, 1: true },
      discountPct: 10,
    });
    const priced = pricePackage(p, { days: 3, nights: 2, durations: DURATIONS, overnightDuration: OVERNIGHT });
    expect(priced.rows).toHaveLength(3);
    // 3 lunches ($35) + 2 overnights ($150) = 105 + 300 = 405; minus 10% = 364.50.
    // (Day 2 & 3 mornings each get a bonus free visit, but the only visit is lunch which
    //  becomes free those days — so lunch is billed on day 1 only.)
    expect(priced.subtotal).toBeGreaterThan(0);
    expect(priced.discount).toBeCloseTo(priced.subtotal * 0.1, 5);
    expect(priced.total).toBeCloseTo(priced.subtotal - priced.discount, 5);
  });
});

describe('gapWarnings', () => {
  it('flags a bare day whose visits leave a gap wider than the max', () => {
    const warns = gapWarnings(
      [{ id: 'a', time: 8 * 60, durationId: 'd1', label: 'AM' }],
      '07:00',
      '22:00',
      4,
      { eveningFrom: null, morningUntil: null, bonusFreeVisit: false },
    );
    expect(warns.length).toBeGreaterThan(0); // 8:00 → 22:00 is 14h, one 8am visit can't cover it
  });
});

describe('quoteText', () => {
  const p = pkg({ name: 'Balanced', visits: [{ id: 'v', time: 12 * 60, durationId: 'd3', label: 'Lunch' }] });
  const priced = { ...pricePackage(p, { days: 2, nights: 1, durations: DURATIONS, overnightDuration: OVERNIGHT }), pkg: p };

  it('renders a client-facing per-day quote', () => {
    const text = quoteText({ clientName: 'Rex', startDate: '2026-07-01', days: 2, priced });
    expect(text).toContain('TribeTails — Coverage Package');
    expect(text).toContain('Prepared for: Rex');
    expect(text).toContain('Balanced · 2 days');
    expect(text).toContain('Lunch (45-min visit)');
    expect(text).toContain('Total');
  });
});

File: auntieos-admin/android/app/src/test/java/com/tribetails/auntieos/domain/CoveragePackageTest.kt
package com.tribetails.auntieos.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Pure-logic tests for the Coverage Package Builder domain (PackageBuilder_7
 * model), mirroring the web `coveragePackage.test.ts` so both platforms are
 * pinned to the same behaviour.
 */
class CoveragePackageTest {

    private val overnight = DEFAULT_DURATIONS.first { it.kind == "overnight" } // d7, $150, 720min

    private fun pkg(
        visits: List<Visit> = emptyList(),
        overnightNights: Map<Int, Boolean> = emptyMap(),
        overnightStart: String = "21:00",
        overnightBufferHours: Double = 2.0,
        discountPct: Double = 0.0,
        discountLabel: String = "",
    ) = Package("p1", "Test", visits, emptyMap(), overnightNights, overnightStart, overnightBufferHours, discountLabel, discountPct)

    @Test
    fun `daysBetween counts inclusively`() {
        assertEquals(5, daysBetween("2026-07-01", "2026-07-05"))
        assertEquals(0, daysBetween("2026-07-05", "2026-07-01"))
    }

    @Test
    fun `withKind migrates the legacy overnight id, defaults the rest to visit`() {
        val migrated = withKind(listOf(Duration("x", "A", 30.0, 20.0, ""), Duration("d7", "Overnight", 720.0, 150.0, "")))
        assertEquals("visit", migrated[0].kind)
        assertEquals("overnight", migrated[1].kind)
    }

    @Test
    fun `durationsFromServiceRates reads the KinCare types as the visit menu`() {
        val rates = mapOf("30Minute" to "25", "Half-Day 6Hrs" to "100", "Overnight" to "150", "Consultation" to "")
        val menu = durationsFromServiceRates(rates).associateBy { it.id }
        // The service NAME is the id, the length comes off the name, the rate off the map.
        assertEquals(Duration("30Minute", "30Minute", 30.0, 25.0, "visit"), menu.getValue("30Minute"))
        assertEquals(360.0, menu.getValue("Half-Day 6Hrs").minutes, 0.001)
        // Named overnight is an overnight; a 6-hour day stay is still a visit.
        assertEquals("overnight", menu.getValue("Overnight").kind)
        assertEquals("visit", menu.getValue("Half-Day 6Hrs").kind)
        // No price and no stated length is 0, not a dropped row.
        assertEquals(0.0, menu.getValue("Consultation").price, 0.001)
        assertEquals(0.0, menu.getValue("Consultation").minutes, 0.001)
    }

    @Test
    fun `durationsFromServiceRates takes a stated duration over the one in the name`() {
        val menu = durationsFromServiceRates(mapOf("Overnight" to "150"), mapOf("Overnight" to "720"))
        assertEquals(720.0, menu[0].minutes, 0.001)
    }

    @Test
    fun `alignPinnedToDurations repoints a pinned visit whose length left the menu`() {
        val menu = durationsFromServiceRates(mapOf("30Minute" to "25", "Overnight" to "150"))
        // The shipped default pins `d2`, an id from the deleted builder-owned menu.
        val aligned = alignPinnedToDurations(DEFAULT_COVERAGE_RULES, menu)
        assertEquals("30Minute", aligned.pinnedTimes[0].durationId)
        assertEquals(DEFAULT_COVERAGE_RULES.pinnedTimes[0].label, aligned.pinnedTimes[0].label)
        assertEquals(DEFAULT_COVERAGE_RULES.pinnedTimes[0].time, aligned.pinnedTimes[0].time)
        // Nothing to point at leaves it blank rather than inventing a length.
        assertEquals("", alignPinnedToDurations(DEFAULT_COVERAGE_RULES, emptyList()).pinnedTimes[0].durationId)
    }

    @Test
    fun `alignPinnedToDurations leaves a pinned visit that already resolves`() {
        val menu = durationsFromServiceRates(mapOf("30Minute" to "25", "60Minute" to "45"))
        val rules = DEFAULT_COVERAGE_RULES.copy(pinnedTimes = listOf(PinnedTime("p", "Meds", "12:00", "60Minute")))
        assertEquals("60Minute", alignPinnedToDurations(rules, menu).pinnedTimes[0].durationId)
    }

    @Test
    fun `todayIso is the device's calendar date`() {
        assertEquals("2026-09-09", todayIso(java.time.LocalDate.of(2026, 9, 9)))
    }

    @Test
    fun `buildDayPatterns is empty for a degenerate window and drops empty suggestions`() {
        assertTrue(buildDayPatterns(DEFAULT_DURATIONS, emptyList(), 6.0, "14:00", "11:00").isEmpty())
        assertTrue(buildDayPatterns(DEFAULT_DURATIONS, emptyList(), 6.0, "11:00", "14:00").isEmpty())
    }

    @Test
    fun `buildDayPatterns only fills with visit-kind durations`() {
        val patterns = buildDayPatterns(DEFAULT_DURATIONS, emptyList(), 3.0, "07:00", "22:00")
        assertTrue(patterns.isNotEmpty())
        for (p in patterns) for (tp in p.touchpoints) {
            assertEquals("visit", DEFAULT_DURATIONS.first { it.id == tp.durationId }.kind)
        }
    }

    @Test
    fun `visitsFromPattern seeds a template`() {
        val p = buildDayPatterns(DEFAULT_DURATIONS, listOf(PinnedTime("p", "Meds", "12:00", "d3")), 12.0, "07:00", "22:00").first()
        val visits = visitsFromPattern(p)
        assertEquals(p.touchpoints.size, visits.size)
        assertEquals(720, visits.first().time)
    }

    @Test
    fun `priceDay bills a bare visit at list price`() {
        val (items, total) = priceDay(listOf(Visit("v", 720, "d3", "Lunch")), DEFAULT_DURATIONS, Coverage(null, null, false))
        assertEquals(35.0, total, 0.001) // d3 = $35
        assertTrue(!items[0].free)
    }

    @Test
    fun `a visit inside the overnight window is covered`() {
        val p = pkg(overnightNights = mapOf(0 to true), overnightStart = "21:00", overnightBufferHours = 2.0)
        val cov = coverageForDay(p, 0, 2, overnight)
        assertEquals(19 * 60, cov.eveningFrom) // 21:00 minus 2h buffer
        val (items, _) = priceDay(listOf(Visit("v", 20 * 60, "d3", "Evening")), DEFAULT_DURATIONS, cov)
        assertTrue(items[0].free)
        assertEquals(0.0, items[0].price, 0.001)
    }

    @Test
    fun `the morning after an overnight grants a bonus free visit`() {
        val cov = Coverage(null, null, true)
        val (items, total) = priceDay(
            listOf(Visit("a", 8 * 60, "d3", "AM"), Visit("b", 12 * 60, "d3", "Noon")),
            DEFAULT_DURATIONS, cov,
        )
        assertTrue(items[0].bonus)
        assertEquals(0.0, items[0].price, 0.001)
        assertEquals(35.0, total, 0.001) // only the second visit is billed
    }

    @Test
    fun `pricePackage applies a discount to the summed days`() {
        val p = pkg(visits = listOf(Visit("v", 12 * 60, "d3", "Lunch")), overnightNights = mapOf(0 to true, 1 to true), discountPct = 10.0)
        val priced = pricePackage(p, PriceContext(3, 2, DEFAULT_DURATIONS, overnight))
        assertEquals(3, priced.rows.size)
        assertTrue(priced.subtotal > 0)
        assertEquals(priced.subtotal * 0.1, priced.discount, 0.001)
        assertEquals(priced.subtotal - priced.discount, priced.total, 0.001)
    }

    @Test
    fun `gapWarnings flags a bare day with a gap wider than the max`() {
        val warns = gapWarnings(listOf(Visit("a", 8 * 60, "d1", "AM")), "07:00", "22:00", 4.0, Coverage(null, null, false))
        assertTrue(warns.isNotEmpty())
    }

    @Test
    fun `quoteText renders a client-facing per-day quote`() {
        val p = pkg(visits = listOf(Visit("v", 12 * 60, "d3", "Lunch"))).copy(name = "Balanced")
        val priced = pricePackage(p, PriceContext(2, 1, DEFAULT_DURATIONS, overnight))
        val text = quoteText(QuoteInput("Rex", "2026-07-01", 2, p, priced))
        assertTrue(text.contains("TribeTails — Coverage Package"))
        assertTrue(text.contains("Prepared for: Rex"))
        assertTrue(text.contains("Balanced · 2 days"))
        assertTrue(text.contains("Lunch (45-min visit)"))
        assertTrue(text.contains("Total"))
    }

    @Test
    fun `minutesToTime formats a 12h label`() {
        assertEquals("7:30 AM", minutesToTime(450.0))
        assertEquals("10:00 PM", minutesToTime(1320.0))
        assertNull(timeToMinutes("nope"))
    }
}
